src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,177 @@
+                return false;
+            }
+
+            await instance.SupplyData(chunk, error);
+            return true;
+        }
+
+        public static async Task<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = Guid.NewGuid();
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            Guid streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _cancellationToken = cancellationToken;
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // TODO: Surely this should be IAsyncEnumerable<ReadOnlySequence<byte>> so we can pass through the
+        // data without having to copy it into a temporary buffer. But trying this gives strange errors -
+        // sometimes the "chunk" variable below has a negative length, even though the logic in BlazorPackHubProtocolWorker
+        // never returns a corrupted item as far as I can tell.
+        private async Task SupplyData(ReadOnlySequence<byte> chunk, string error)
+        {
+            if (chunk.Length < 0)
+            {
+                throw new InvalidOperationException($"The incoming data chunk cannot be negative.");
+            }
+
+            try
+            {
+                if (!string.IsNullOrEmpty(error))
+                {
+                    throw new InvalidOperationException($"An error occurred while reading the remote stream: {error}");
+                }
+
+                if (chunk.Length == 0)
+                {
+                    throw new InvalidOperationException($"The incoming data chunk cannot be empty.");
+                }
+
+                _bytesRead += chunk.Length;
+
+                if (_bytesRead > _totalLength)
+                {
+                    throw new InvalidOperationException($"The incoming data stream declared a length {_totalLength}, but {_bytesRead} bytes were read.");
+                }
+
+                // Enforce 1 MB max chunk size.
+                const int maxChunkLength = 1024 * 1024;
```

Comment:
Should we be checking this? The SignalR max message size limit already covers what we want, doesn't it? There's no need to enforce a limit in the wasm case.

(BTW I know this came from my proof-of-concept)